artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageCursorProviderImpl.java

```diff
@@ -132,43 +156,88 @@ public PagedReference newReference(final PagePosition pos,
    @Override
    public PageCache getPageCache(final long pageId) {
       try {
+         if (pageId > pagingStore.getCurrentWritingPage()) {
+            return null;
+         }
+         boolean createPage = false;
+         CompletableFuture<PageCache> inProgressReadPage;
          PageCache cache;
+         Page page = null;
          synchronized (softCache) {
-            if (pageId > pagingStore.getCurrentWritingPage()) {
+            cache = softCache.get(pageId);
+            if (cache != null) {
+               return cache;
+            }
+            if (!pagingStore.checkPageFileExists((int) pageId)) {
                return null;
             }
-
-            cache = softCache.get(pageId);
-            if (cache == null) {
-               if (!pagingStore.checkPageFileExists((int) pageId)) {
-                  return null;
-               }
-
+            inProgressReadPage = inProgressReadPages.get(pageId);
+            if (inProgressReadPage == null) {
+               final CompletableFuture<PageCache> readPage = new CompletableFuture<>();
                cache = createPageCache(pageId);
-               // anyone reading from this cache will have to wait reading to finish first
-               // we also want only one thread reading this cache
-               logger.tracef("adding pageCache pageNr=%d into cursor = %s", pageId, this.pagingStore.getAddress());
-               readPage((int) pageId, cache);
-               softCache.put(pageId, cache);
+               page = pagingStore.createPage((int) pageId);
+               createPage = true;
+               inProgressReadPage = readPage;
+               inProgressReadPages.put(pageId, readPage);
             }
          }
+         //We need to create a new criticalComponent each time, because
+         //each read page action can happen from a different thread
+         final CriticalComponent readPageActionComponent;
+         if (criticalAnalyzer == null) {
+            readPageActionComponent = null;
+         } else {
+            final SimpleString address = pagingStore.getAddress();
+            readPageActionComponent = new CriticalComponentImpl(criticalAnalyzer, PAGE_CURSOR_PROVIDER_CRITICAL_PATHS) {
```

Comment:
Do we need to cache the CriticalComponentImpl here to reduce memory footprint?

Reply:
I have no idea, maybe yes, but I don't want to make this code more complex then necessary: consider that on each `Page::read` request we already create 1 `CompleableFuture` and each completed read will produce several Paged messages as well: saving just 1 instance of CriticalComponentImpl maybe is not a super-win...
I'm more concerned from a functional point of view instead:  in this version I'm triggering the critical analyzer in both cases ie on Page::read and on who is waiting for it. 
Probably I can just leave the critical analyzer to monitor Page:.read and just logging otherwise.. @clebertsuconic wdyt?

Reply:
I would expect the actual component to implement CriticalComponent

Reply:
@franz1981 the implementation here is wrong I think..

you should have PageCursorProviderImpl extending CriticalComponentImpl.

Add it upon construction.. (check remove points when the address is removed)

and just call enter / leave...


like all the other implementations.

Reply:
I wasn't sure about it because you can have several concurrent enterCritical() calls on the same component,while I want the critical analyser to report which pageId read was too slow... 
Wdyt?

Reply:
It wasn't engineered for that.. it would be too much information.

if the Critical Analyzer kicks in (remember, it wasn't supposed to), the relevant information is the stack trace and the methods where you have a possible deadlock.

Reply:
I was thinking to let Page extends CriticalComponentImpl and wrap the read with enter/exit..leaving the CompleteableFuture without it (it is not needed we monitor page read), given that we don't need to monitor both (will leave just the timeout logs).
I know that is not a perfect fit for the critical analyser ..but it could work?

Reply:
@franz1981 you are reinventing things on the CriticalAnalyzer. I wouldn't go that route. Just stick to what's already there.

The CriticalAnalyzer is already complicated enough :)

The CriticalAnalyzer was intended to capture deadlocks. upon a critical call you call enter / leave on a finalize call. I strongly recommend against anything different.

it could work? I don't know.. you're using it in a different way than originally written, and I would not have enough time to test it.

Reply:
The original design expected the component on the list as long as it lives. you are for instance adding and removing it like crazy on each write. I don't have an idea of what would be the implications. Definitely not being used as intended.

Reply:
Got it , but given they concurrent behaviour of such component if you have 1 slow page read and all the others fast enough, the slowest one won't be detected..
I have already a test in this PR to detect slow page read without concurrenct requests: I will create a new one just to check if my guess is correct, thanks!

Reply:
@franz1981 yes, you're right... it won't work with multiple Threads.. I don't know what to do on that case.

perhaps we need multiple Components.. but I would on that case use a ThreadLocal or some other cache. I don't know.

Reply:
Perhaps implement something on the PageCursor where isExpired would return false, and have individual measures using some sort of thread local?